### PR TITLE
feat: reclaim old task records from the download status file

### PR DESCRIPTION
**Pull request by phil-IO-p**
## What this does

Bounds the growth of the per-queue download-status file, which until now only ever grew. `_build_status_file_content` merges each `deadline queue sync-output` run onto the previous file and nothing was ever removed, so size scaled with the total number of frames ever downloaded through a queue.

Once more than **100,000 task records** are retained, the records of the least recently updated eligible jobs are emptied. Job rows themselves are never removed.

Counterpart change on the reader side, a 64 MiB read ceiling in Deadline Cloud Monitor.

## Why the budget counts frames, not jobs

Task records dominate size. Measured on the pretty-printed file, two independent measurements agreeing within 3%:

| Item | Bytes |
|---|---|
| One task record | 230 to 236 |
| One job row | 329 to 406 |

Capping job count instead would leave size at the mercy of frames-per-job:

| Frames per job | File size under "newest 50 jobs" |
|---|---|
| 100 | 1 MB |
| 1,000 | 11.5 MB |
| 10,000 | **115 MB** |

That last row fails worst for the large-render studios the feature exists to serve.

## What is protected, and why

Two sets of records are never reclaimed.

**Jobs that did not finish cleanly.** A job is eligible only if every task record is a plain success, with no `failed`, no `farm_failed`, and no `error_code`. The Monitor derives a job's partial-failure indicator from the records alone rather than from the job's counts, so dropping them from a partly-failed job would repaint it as a clean success. Job status alone is not sufficient here: a `downloaded` job can legitimately hold failed tasks.

**This run's download candidates.** The `farm_failed` guard in this module reads records from disk to suppress a failure the API replays after a requeue, so emptying them for a job being actively synced would let an already-downloaded frame be recorded as a farm failure.

That second protection deliberately covers the candidate categories only, **not** every job the checkpoint knows about. An inactive job is re-appended to the checkpoint on every run so a later requeue can be detected (`_incremental_download.py:1201`), and `inactive` is recomputed as "in the checkpoint but not a candidate", so a finished job is inactive permanently. Protecting that set would exempt exactly the accumulated history this change exists to reclaim, and the overshoot branch would become the steady state rather than the exception.

## Accepted consequences

- **The file is not bounded.** Job rows accumulate at ~330 bytes each, roughly 33 MB per 100,000 jobs. Only records are reclaimed. See the next section for why rows are left alone.
- **The record target is a reclaim budget, not a ceiling.** Records held by an ineligible job, or by a candidate this run, are kept without spending it, so the retained total can exceed the target. The overshoot is logged at `info` with both figures broken out, so an operator can tell protected records from unreclaimed ones.
- **A reclaimed job that is later requeued can misreport one frame.** With no on-disk record for the `farm_failed` guard to match, a frame whose output was downloaded earlier can be recorded as a farm failure. The same applies to a job another machine is tracking, since one machine cannot see another's candidate set. This is the price of the change reclaiming anything at all; the alternatives were a no-op or zeroed file counts.
- **A queue whose jobs mostly carry failures stays large.** Little is eligible there. Lifting this needs a marker on reclaimed jobs plus new logic in the merge path, left as a follow-up.
- **The first run after this ships still parses the old oversized file in full** inside the lock, since the read precedes any reclaim. Measured at 367 ms for a 101 MB file. One-time, and the file it writes back is smaller.

## Why job rows are never removed

Deleting a row makes the next run rebuild it from scratch with zeroed file counts, because the preserve-counts branch needs the existing entry to be there. Avoiding that requires knowing a job is finished with, and nothing in this file can show it:

- `last_updated` moves only on a status change, never on a sync. A job that finished weeks ago and is still enumerated every run carries an old timestamp while being actively tracked.
- Several machines write this file with independent checkpoints, so one machine's job set proves nothing about what another still tracks.

An age gate on `last_updated` was tried and removed for that reason: it reads "nothing changed recently" as "safe to delete", which is the normal steady state of a healthy finished job. Bounding rows needs a field that advances on every sync. That would also lift the eligibility restriction above, so it belongs in its own change rather than bolted on here.

## Shared-filesystem hazards handled

This file is written by several machines to a shared location, so the retention order cannot trust its contents:

- **A timestamp further ahead than a five-minute skew allowance is clamped to now**, not demoted to oldest. A machine whose clock is fast is writing its newest entries, so treating them as oldest would make every other writer reclaim that machine's records first. The allowance is one-sided by nature: a clock reading in the past produces timestamps indistinguishable from genuinely old ones.
- **`fromisoformat` only accepts a trailing `Z` from Python 3.11**, and `requires-python` is `>=3.9`, so which records were reclaimed first would otherwise depend on the interpreter running the sync. `Z` is now normalised.
- **Entries and records of an unexpected JSON shape are treated as ineligible instead of raising.** `write_download_status_file` swallows every exception and only logs, so a shape error in the reclaim pass would have left that queue's file frozen at its old content indefinitely.

## Why `feat:` and not `feat!:`

Deliberate, and worth a reviewer's disagreement. The schema is unchanged, no CLI subcommand, argument or default changes, and no job row ever disappears. What does change is that an empty `tasks` dict can now mean "records reclaimed" as well as "nothing downloaded", which is called out on the schema-version constant.

## Testing

- `hatch run test`: **3842 passed, 22 skipped, 0 failures**. Coverage 77.67% against the 69% gate; the changed module is at 95%, with the remaining uncovered lines pre-existing.
- `hatch run fmt`, `hatch run lint` (ruff + mypy), `hatch build`: all clean.
- 21 new tests plus 7 parametrized cases, including one that goes through the public `write_download_status_file` path with a real read/write cycle rather than only the internal builder.
- **Every guard was verified by removing it and confirming its test fails.** That check caught three tests of mine that passed for reasons unrelated to what they were written to check, including one whose subject was unreachable because a recently-settled row sorts to the front rather than into the tail.
- Not run: the integration suite. Nothing was measured against a real network share, so any timing figure here is CPU only.

### CODE REVIEW

File: src/deadline/client/cli/_download_status_file.py [hunk 1]
```diff
@@ -233,6 +245,128 @@ def _determine_job_download_status(
...
+        elif budget > 0:
+            budget = max(0, budget - record_count)
+            retained_records += record_count
+        else:
+            entry["tasks"] = {}
+            dropped_records += record_count
+
+    dropped_rows = 0
+    for _, (job_id, entry) in ordered[_TARGET_RETAINED_JOB_ROWS:]:
+        if job_id not in current_run_job_ids and _is_history_trimmable(entry):
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: Good catch, and the multi-writer point is the part I had wrong. Fixed by taking your suggestion: a row now has to be settled for 30 days (`_MIN_ROW_AGE_BEFORE_DELETE`) as well as absent from this run, since one machine's job set proves nothing about what another machine still tracks.

I have left the residual stated rather than claimed closed: a job another machine has tracked for longer than that age, without its status changing in that time, can still be trimmed. The docstring says so.

New test `test_row_is_not_deleted_until_its_status_has_been_settled_long_enough`. Worth noting it took two attempts: my first version put a recently-settled row in the file, but a recent row sorts to the front and so was never in the deletable tail, and the test passed with the guard removed. It now stamps every row recently so all of them are past the row cap by count.

File: src/deadline/client/cli/_download_status_file.py [hunk 2]
```diff
@@ -233,6 +245,128 @@ def _determine_job_download_status(
...
+        record_count = len(tasks) if isinstance(tasks, dict) else 0
+        if not record_count:
+            continue
+        if not _is_history_trimmable(entry):
+            retained_records += record_count
+        elif budget > 0:
+            budget = max(0, budget - record_count)
+            retained_records += record_count
+        else:
+            entry["tasks"] = {}
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py:362]: Correct, and this is the more interesting of the two. You are right that the docstring claim was backwards for this case: the guard reads the `downloaded` records, so an all-clean job is exactly the one whose trimming can produce a wrong record after a requeue. I had reasoned only about failures that exist at trim time and missed the requeue path entirely.

Taken the simplest fix you suggested: the record loop now skips `job_id in current_run_job_ids` too, and the docstring says why rather than asserting the old claim. New test `test_records_of_a_job_in_the_current_run_are_never_emptied`, verified to fail when the exclusion is removed.

File: src/deadline/client/cli/_download_status_file.py [hunk 3]
```diff
@@ -233,6 +245,128 @@ def _determine_job_download_status(
...
+
+    if dropped_records or dropped_rows:
+        logger.debug(
+            f"Trimmed download status history: dropped {dropped_records} task records and "
+            f"{dropped_rows} job rows, retaining {retained_records} records."
+        )
+    if retained_records > _TARGET_RETAINED_TASK_RECORDS:
+        logger.warning(
+            f"Download status file retains {retained_records} task records, above the "
+            f"{_TARGET_RETAINED_TASK_RECORDS} target, because jobs with failed frames are never "
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: Agreed on both counts. The message asserted one cause when there are three, and two of them are by design.

Now logged at `info`, stating the retained count, the target, and how many of those records belong to jobs that are not eligible to be trimmed, without claiming why. That keeps the number visible for anyone debugging a slow queue without telling them to go looking for failures that may not exist.

File: src/deadline/client/cli/_download_status_file.py [hunk 4]
```diff
@@ -233,6 +245,128 @@ def _determine_job_download_status(
...
+    try:
+        # fromisoformat only accepts a trailing "Z" from 3.11, and this file is written by
+        # whatever client reached it first.
+        parsed = datetime.fromisoformat(raw[:-1] + "+00:00" if raw.endswith("Z") else raw)
+    except ValueError:
+        return _OLDEST_SORT_KEY
+    if parsed.tzinfo is None:
+        parsed = parsed.replace(tzinfo=timezone.utc)
+    if parsed > datetime.now(timezone.utc) + _CLOCK_SKEW_ALLOWANCE:
+        return _OLDEST_SORT_KEY
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: You are right and this reverses what I had. I had moved from clamping to demoting because a clamped future timestamp still outranks a genuinely older entry, but that traded a small ordering inaccuracy for actively destroying the history of whichever machine has a bad clock, which is much worse. Adopted your version.

Also dropped the lock-TTL analogy from the comment justifying the five minutes; you are right that a liveness timeout on a shared file says nothing about clock agreement. The comment now just says what the allowance is for and that being wrong either way costs history, not correctness.

The test flipped with it, and now asserts the property you describe: a skewed entry is not the first thing evicted.

File: src/deadline/client/cli/_download_status_file.py [hunk 5]
```diff
@@ -10,15 +10,25 @@
...
 
 from ..config.config_file import DEFAULT_QUEUE_INCREMENTAL_DOWNLOAD_DIR
 from ._incremental_download import CategorizedJobIds
 
 logger = logging.getLogger(__name__)
 
+# Covers the file's shape, which trimming does not change. A missing job row or an empty tasks
+# dict therefore means "trimmed" as readily as "never synced"; readers must not treat an absent
+# entry as evidence a job was never downloaded.
 DOWNLOAD_STATUS_FILE_SCHEMA_VERSION = 1
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py:23]: Confirmed on the reader side rather than assumed, since the same team owns the Monitor. Two new states are reachable, and they are not equally bad:

A retained row with an empty `tasks` dict is safe. The partial-failure indicator counts only records whose status is a failure, so an empty dict yields zero failures, which is the truth for an eligible job: eligibility requires every record to be an explicit success. The progress denominator prefers the job API's live task-run counts and falls back to the record count only when those are unavailable. So the column renders the job as fully downloaded, which is correct.

An absent row is the real loss. It renders as the empty cell, indistinguishable from a job that was never synced, so a customer cannot tell "history was trimmed" from "this never downloaded". That is the accepted cost. It is now much rarer than in the version you reviewed, since a row also has to have been settled for 30 days and absent from the current run.

Kept at version 1 deliberately. Bumping it would make every already-deployed reader reject the whole file as an unsupported schema and show nothing for the queue, which is strictly worse than the ambiguity above. A marker on trimmed entries is the better shape for this, and it is the natural follow-up if the ambiguity turns out to matter, since it also lifts the eligibility restriction that keeps failure-heavy queues growing. I did not want to add a field and new merge-path logic in the same change that introduces trimming.

File: src/deadline/client/cli/_download_status_file.py [hunk 6]
```diff
@@ -233,6 +245,155 @@ def _determine_job_download_status(
...
+            retained_records += record_count
+        else:
+            entry["tasks"] = {}
+            dropped_records += record_count
+
+    dropped_rows = 0
+    for last_updated, _, job_id, entry in ordered[_TARGET_RETAINED_JOB_ROWS:]:
+        if (
+            job_id not in current_run_job_ids
+            and last_updated < now - _MIN_ROW_AGE_BEFORE_DELETE
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: Real bug, and the sharpest of the three. You are right that `datetime.min < now - 30 days` is trivially true, so the gate was satisfiable with no age evidence at all, and it hit exactly the entries the tolerant shape handling exists for.

Fixed close to your suggestion, though one step further than the identity check: `_last_updated_sort_key` is now `_last_updated`, returning `Optional[datetime]`, so "cannot date this" and "dated long ago" are different values rather than the same sentinel. Ordering maps `None` to the sentinel at the sort call; row deletion requires `last_updated is not None`. An undatable row now keeps its row forever and only loses its records, which is the direction the design already prefers.

New test `test_undatable_row_is_never_deleted`, verified to fail when the `is not None` check is removed. It also asserts the gate still fired for the datable rows, so it cannot pass by the cap simply not triggering.

Agreed on the backwards-skew asymmetry too. Nothing here can distinguish a slow clock from genuinely old data, so requiring positive age evidence is the only defence, and that is now what the code does.

File: src/deadline/client/cli/_download_status_file.py [hunk 7]
```diff
@@ -233,6 +245,155 @@ def _determine_job_download_status(
...
+        parsed = datetime.fromisoformat(raw[:-1] + "+00:00" if raw.endswith("Z") else raw)
+    except ValueError:
+        return _OLDEST_SORT_KEY
+    if parsed.tzinfo is None:
+        parsed = parsed.replace(tzinfo=timezone.utc)
+    now = datetime.now(timezone.utc)
+    # Clamped rather than demoted to oldest. A machine whose clock is hours fast is writing its
+    # newest entries, so treating them as the oldest would make every other writer destroy that
+    # machine's history first.
+    return min(parsed, now) if parsed > now + _CLOCK_SKEW_ALLOWANCE else parsed
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py:303]: Correct, the docstring was stale from the demote-to-oldest version and the inline comment described the shipped behaviour. Rewritten so the function says what it does: the timestamp is clamped, with the residual that a fast-clock machine wins retention ties and so has its history preferentially kept.

The one-sidedness is now stated where it matters rather than left implicit: a clock reading in the past is indistinguishable from genuinely old data, which is why row deletion requires positive age evidence instead of trusting the order.

File: src/deadline/client/cli/_download_status_file.py [hunk 8]
```diff
@@ -233,6 +245,155 @@ def _determine_job_download_status(
...
+            retained_records += record_count
+        elif budget > 0:
+            budget = max(0, budget - record_count)
+            retained_records += record_count
+        else:
+            entry["tasks"] = {}
+            dropped_records += record_count
+
+    dropped_rows = 0
+    for last_updated, _, job_id, entry in ordered[_TARGET_RETAINED_JOB_ROWS:]:
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: Both parts correct, and the counter bug was straightforwardly wrong: the figures described a file that was never written.

On the counters: the row pass now runs first and the record loop skips job ids no longer in `jobs_status`, so records belonging to a deleted row land in neither figure. Covered by `test_logged_counts_describe_the_file_that_gets_written`. Getting that test to actually catch the bug took a second attempt: my first version sized the records so the deleted row fell into the dropped bucket either way, and it passed with the guard removed. With counts small enough to stay inside the budget, the buggy version reports 8 retained records for a file containing 6.

On the interaction you identified, I took your second direction rather than the first. Waiving the age gate at a hard maximum reintroduces exactly the zeroed-counts rebuild this change added the gate to prevent, and I would rather the file be larger than wrong. So the row figure is now documented as a cleanup threshold rather than a ceiling, records are called out as the only hard-bounded dimension, and there is a second `logger.info` reporting row overshoot with the reason, since as you say an operator at 10x the intended row count previously got no signal at all.

That does mean this change does not bound row count for a queue syncing more than 2,000 distinct jobs inside the window. Stated plainly in the docstring and the commit message rather than implied by the constant name. Closing it properly needs a marker on trimmed entries, which would also lift the eligibility restriction that keeps failure-heavy queues growing, and I did not want to add a field and new merge-path logic in the same change that introduces trimming.

File: src/deadline/client/cli/_download_status_file.py [hunk 9]
```diff
@@ -233,6 +242,145 @@ def _determine_job_download_status(
...
+        key=lambda item: (item[0] or _OLDEST_SORT_KEY, item[1]),
+        reverse=True,
+    )
+
+    dropped_rows = 0
+    for last_updated, _, job_id, entry in ordered[_TARGET_RETAINED_JOB_ROWS:]:
+        if (
+            job_id not in current_run_job_ids
+            and last_updated is not None
+            and last_updated < now - _MIN_ROW_AGE_BEFORE_DELETE
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: You are right, and this one goes to the root: the gate rests on a field whose contract is that it stands still. `last_updated` moving only on a status change means a healthy finished job that is still enumerated carries an old timestamp indefinitely, so the gate reads "nothing changed recently" as "safe to delete". I had been treating those as the same thing and they are not.

Rather than reach for a `last_seen`, I have removed row deletion from this change entirely. Records-only reclaim, with the current-run exclusion. The zeroed-counts rebuild is now unreachable in every configuration, single machine or multi, because no row is ever removed. `_MIN_ROW_AGE_BEFORE_DELETE` and the row target are gone, and six tests went with them.

The cost is stated plainly rather than implied: the file is no longer bounded. Rows accumulate at ~330 bytes each, about 33MB per 100,000 jobs. Bounding them needs exactly what you describe, a field that advances on every sync, and that also lifts the eligibility restriction keeping failure-heavy queues large, so it deserves its own change and its own review rather than being added at round four of this one.

The residual you name in the last paragraph survives and is now the only cross-machine hazard: another machine can empty the records of a job this one is tracking, so a later requeue could record an already-downloaded frame as a farm failure. It is in the docstring. It is narrower than what it replaces, but it is real and I have not claimed otherwise.

File: src/deadline/client/cli/_download_status_file.py [hunk 10]
```diff
@@ -233,6 +242,145 @@ def _determine_job_download_status(
...
+
+    Task records are dropped whole per job, since a half-populated tasks dict understates a job's
+    frame count. Ties in last_updated fall back to file position, latest first.
+
+    Nothing this run touched is trimmed: the farm_failed guard in _build_status_file_content reads
+    the on-disk records, and its preserve-counts branch needs the existing row.
+
+    Deleting a row additionally needs a readable timestamp and a status unchanged for
+    _MIN_ROW_AGE_BEFORE_DELETE, because other machines write this file with independent checkpoints
+    and this run's job set says nothing about what they still track. Records are the only
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: Both correct. "Hard-bounded" was wrong on its own terms, and the branch three lines down proves it.

The docstring now says the target is a reclaim budget rather than a ceiling, and names the two kinds of record that are kept without spending it. With row deletion gone there is no remaining claim that anything is bounded.

The log accuracy problem is fixed the way you suggest, as two numbers rather than one: it reports how many retained records belong to jobs that are not eligible and how many to jobs the run is still tracking. The case you describe, where the overshoot is entirely live jobs and the line reported "0 not eligible", would now read as protected rather than as the trimmer failing.

File: src/deadline/client/cli/_download_status_file.py [hunk 11]
```diff
@@ -233,6 +242,145 @@ def _determine_job_download_status(
     return _make_status_entry("in_progress", total_files, downloaded_files)
 
 
+_OLDEST_SORT_KEY = datetime.min.replace(tzinfo=timezone.utc)
+# Beyond this, a timestamp is a clock error rather than a newer entry. Affects order only.
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py:244]: "Affects order only" was false when you read it and is true now: with row deletion gone, the clamped value has no consumer other than the sort, so the un-reclaimable-row problem and the never-persisted-correction problem both stop existing. `_last_updated` is back to a plain sort key returning the oldest sentinel for undatable entries, since the distinction it existed to carry had only one consumer and that consumer is gone.

I did not take the re-stamping suggestion. Writing a corrected `last_updated` back would have this machine overwrite another machine's record of when a job last changed, on the strength of a clock comparison, which is the same class of cross-machine assumption that caused the finding above. Worth revisiting alongside a field that actually tracks sightings.

File: src/deadline/client/cli/_download_status_file.py [hunk 12]
```diff
@@ -348,6 +463,9 @@ def _build_status_file_content(
             existing_entry["last_updated"] = now
             _reconcile_entry_with_tasks(existing_entry)
 
+    # Ordered after reconciliation, which reads entry["tasks"].
+    _trim_retained_history(jobs_status, all_job_ids | categorized_job_ids.inactive)
```
[the remark this answers is not in the document]
Comment (phil-IO-p) [reply src/deadline/client/cli/_download_status_file.py]: Verified against `_incremental_download.py` rather than taken on trust, and you are right on every step. Line 1201 re-appends `IncrementalDownloadJob({"jobId": job_id}, session_ended_timestamp, {})` for each inactive job on every run, `job_session_ended_timestamps` is seeded from `checkpoint.jobs` where the timestamp is not None so the placeholder reproduces itself, and `inactive` is recomputed at line 419 as `checkpoint_job_ids - download_candidate_job_ids`. A finished job is therefore inactive permanently, and protecting that set exempted exactly the accumulated history this change exists to reclaim. The overshoot branch would have been the steady state.

Fixed by protecting `all_job_ids` only, as you suggest. New test `test_records_of_an_inactive_job_are_reclaimed`, and I confirmed it fails against the previous `all_job_ids | categorized_job_ids.inactive`, so it pins the behaviour rather than just describing it.

You are also right about why my tests missed it: `_build_from_history` passed an empty `inactive` set, a state the real caller never produces for a job with sessions. That is the second time in this review a test of mine passed for a reason unrelated to the thing it was written to check, so I have started verifying each new guard by removing it and confirming the test fails.

On the residual: I have taken it rather than argued it away. A job whose records were reclaimed and which is later requeued can have an already-downloaded frame recorded as a farm failure, since the guard has nothing on disk to match. That is now stated in the docstring in one clause and in the commit message at length. It is the price of the change doing anything at all, and it is narrower than the alternatives: protecting inactive jobs makes this a no-op, and deleting rows produced zeroed file counts.

File: src/deadline/client/cli/_download_status_file.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (larrygao001) [reply src/deadline/client/cli/_download_status_file.py:293]: We have uncommited changes for this?

File: src/deadline/client/cli/_download_status_file.py [hunk 13]
```diff
@@ -233,6 +240,112 @@ def _determine_job_download_status(
...
+            retained_records += record_count
+        else:
+            entry["tasks"] = {}
+            dropped_records += record_count
+
+    if dropped_records:
+        logger.debug(
+            f"Trimmed download status history: dropped {dropped_records} task records, "
+            f"retaining {retained_records}."
+        )
```
Comment (larrygao001) [inline src/deadline/client/cli/_download_status_file.py:340]: nit: can we make this more visible to the users?

File: src/deadline/client/cli/_download_status_file.py [hunk 14]
```diff
@@ -233,6 +240,112 @@ def _determine_job_download_status(
...
+    if retained_records > _TARGET_RETAINED_TASK_RECORDS:
+        logger.info(
+            f"Download status file retains {retained_records} task records, above the "
+            f"{_TARGET_RETAINED_TASK_RECORDS} target: {untrimmable_records} belong to jobs that "
+            f"are not eligible to be trimmed and {live_records} to jobs this run is still tracking."
```
Comment (larrygao001) [inline src/deadline/client/cli/_download_status_file.py:345]: Can we make this more visible to the users?

**Review by larrygao001 — APPROVED** [review]

**Comment by andychoquette** [comment]
In instances where a task row has been reclaimed, what is the customer experience. Are those task rows stale until synced? I'm not super familiar with this functionality.

**Comment by phil-IO-p** [comment]
> In instances where a task row has been reclaimed, what is the customer experience. Are those task rows stale until synced? I'm not super familiar with this functionality.

@andychoquette  Good question, and it turned up something we had missed, so thanks for asking!

**Short answer**
They do not come back on their own. A sync only writes per-frame records for frames it downloads on that run, and a job that already finished and downloaded is not re-downloaded, so once its per-frame detail is reclaimed it stays gone. "Stale until synced" would actually be the better outcome; the reality is closer to "gone unless something forces a re-download of that job".

**What a customer would see today, for a job whose per-frame detail has been reclaimed**

- The job-level Download status stays correct. The row keeps its status and its file counts, so the job still reads as Downloaded with the right totals.
- The per-frame Download status column in the job panel is where the problem is. With no per-frame record, and the job still present in the file, the column concludes that the sync looked and found nothing, and shows `No outputs` for every frame. Those outputs were downloaded and are on disk, so that is simply wrong, and it is worse than showing a dash because it makes a positive claim.

The underlying reason is that the reader cannot currently tell "these records were reclaimed to keep the file small" apart from "this job never had records", and those two cases deserve different labels.

So I am holding this as a draft rather than pushing it forward. Landing it safely needs two things together:
1. A marker in the status file saying a job's per-frame detail was intentionally reclaimed, and
2. A change on the reader side to honor that marker instead of concluding `No outputs`. I had deliberately avoided adding a field to the file, but this is the third separate problem that points at needing one, so that looks like the right call now.

**Comment by phil-IO-p** [comment]
Closing this. Clearing a finished job's task records makes Deadline Cloud Monitor report downloaded frames as having no outputs, and deleting the whole row breaks the same way as soon as the job is requeued.

Replaced by #1361: write the file compactly (40% smaller) and report its size, so any ceiling comes from real numbers instead of a guess.
